### ISSUE TITLE
retry DAG workflow with depends failed with msg ` Ancestor task node step2 not found`

### ISSUE THREAD

**Issue by heidongxianhua** [argoproj/argo-workflows#12924]
### Pre-requisites

- [X] I have double-checked my configuration
- [X] I can confirm the issue exists when I tested with `:latest`
- [X] I have searched existing issues and could not find a match for this bug
- [ ] I'd like to contribute the fix myself (see [contributing guide](https://github.com/argoproj/argo-workflows/blob/main/docs/CONTRIBUTING.md))

### What happened/what did you expect to happen?

retry succeed
failed wf 
![image](https://github.com/argoproj/argo-workflows/assets/22924371/e2a237a6-cff1-4fb7-baab-b49dfabe3f1f)
 
and then i retry 

![image](https://github.com/argoproj/argo-workflows/assets/22924371/45365329-a7e5-414a-9c02-b20ea88510b8)

### Version

lastest

### Paste a small workflow that reproduces the issue. We must be able to run the workflow; don't enter a workflows that uses private images.

```YAML
apiVersion: argoproj.io/v1alpha1
kind: Workflow
metadata:
  labels:
    workflows.argoproj.io/creator: system-serviceaccount-argo-argo-server
  name: dag-test
  namespace: argo
spec:
  entrypoint: main
  templates:
  - name: main
    dag:
      tasks:
        - name: step1
          template: run-pod-1
        - name: step2
          template: run-pod-2
        - name: step3
          depends: step1.Succeeded && (step2 || step2.Failed)
          template: run-pod-1
        - name: step4
          depends: step3.Succeeded
          template: run-pod-2
  - name: run-pod-1
    container:
      image: docker/whalesay:latest
      command: [python3, -c]
      args: ["print('gfn')"]
  - name: run-pod-2
    container:
      image: docker/whalesay:latest
      command: [python3, -c]
      args: ["print('gfn')\nraise ValueError('raise value error')"]
  ttlStrategy:
    secondsAfterCompletion: 3600
  podGC:
    strategy: OnPodCompletion
```

### Logs from the workflow controller

```text
time="2024-04-11T06:44:24.035Z" level=info msg="Processing workflow" namespace=argo workflow=dag-test-t9l2s
time="2024-04-11T06:44:24.051Z" level=info msg="Updated phase  -> Running" namespace=argo workflow=dag-test-t9l2s
time="2024-04-11T06:44:24.058Z" level=warning msg="Node was nil, will be initialized as type Skipped" namespace=argo workflow=dag-test-t9l2s
time="2024-04-11T06:44:24.058Z" level=info msg="was unable to obtain node for , letting display name to be nodeName" namespace=argo workflow=dag-test-t9l2s
time="2024-04-11T06:44:24.058Z" level=info msg="DAG node dag-test-t9l2s initialized Running" namespace=argo workflow=dag-test-t9l2s
time="2024-04-11T06:44:24.058Z" level=warning msg="was unable to obtain the node for dag-test-t9l2s-4187141573, taskName step4"
time="2024-04-11T06:44:24.058Z" level=warning msg="was unable to obtain the node for dag-test-t9l2s-4069698240, taskName step3"
time="2024-04-11T06:44:24.058Z" level=warning msg="was unable to obtain the node for dag-test-t9l2s-4086475859, taskName step2"
time="2024-04-11T06:44:24.058Z" level=warning msg="was unable to obtain the node for dag-test-t9l2s-4086475859, taskName step2"
time="2024-04-11T06:44:24.058Z" level=info msg="All of node dag-test-t9l2s.step2 dependencies [] completed" namespace=argo workflow=dag-test-t9l2s
time="2024-04-11T06:44:24.058Z" level=warning msg="Node was nil, will be initialized as type Skipped" namespace=argo workflow=dag-test-t9l2s
time="2024-04-11T06:44:24.058Z" level=info msg="Pod node dag-test-t9l2s-4086475859 initialized Pending" namespace=argo workflow=dag-test-t9l2s
time="2024-04-11T06:44:24.084Z" level=info msg="Created pod: dag-test-t9l2s.step2 (dag-test-t9l2s-run-pod-2-4086475859)" namespace=argo workflow=dag-test-t9l2s
time="2024-04-11T06:44:24.084Z" level=warning msg="was unable to obtain the node for dag-test-t9l2s-4103253478, taskName step1"
time="2024-04-11T06:44:24.084Z" level=warning msg="was unable to obtain the node for dag-test-t9l2s-4103253478, taskName step1"
time="2024-04-11T06:44:24.084Z" level=info msg="All of node dag-test-t9l2s.step1 dependencies [] completed" namespace=argo workflow=dag-test-t9l2s
time="2024-04-11T06:44:24.084Z" level=warning msg="Node was nil, will be initialized as type Skipped" namespace=argo workflow=dag-test-t9l2s
time="2024-04-11T06:44:24.084Z" level=info msg="Pod node dag-test-t9l2s-4103253478 initialized Pending" namespace=argo workflow=dag-test-t9l2s
time="2024-04-11T06:44:24.101Z" level=info msg="Created pod: dag-test-t9l2s.step1 (dag-test-t9l2s-run-pod-1-4103253478)" namespace=argo workflow=dag-test-t9l2s
time="2024-04-11T06:44:24.101Z" level=warning msg="was unable to obtain the node for dag-test-t9l2s-4069698240, taskName step3"
time="2024-04-11T06:44:24.101Z" level=warning msg="was unable to obtain the node for dag-test-t9l2s-4187141573, taskName step4"
time="2024-04-11T06:44:24.101Z" level=warning msg="was unable to obtain the node for dag-test-t9l2s-4069698240, taskName step3"
time="2024-04-11T06:44:24.101Z" level=warning msg="was unable to obtain the node for dag-test-t9l2s-4187141573, taskName step4"
time="2024-04-11T06:44:24.101Z" level=info msg="TaskSet Reconciliation" namespace=argo workflow=dag-test-t9l2s
time="2024-04-11T06:44:24.101Z" level=info msg=reconcileAgentPod namespace=argo workflow=dag-test-t9l2s
time="2024-04-11T06:44:24.108Z" level=info msg="Workflow update successful" namespace=argo phase=Running resourceVersion=15178675 workflow=dag-test-t9l2s
time="2024-04-11T06:44:34.036Z" level=info msg="Processing workflow" namespace=argo workflow=dag-test-t9l2s
time="2024-04-11T06:44:34.036Z" level=info msg="Task-result reconciliation" namespace=argo numObjs=1 workflow=dag-test-t9l2s
time="2024-04-11T06:44:34.036Z" level=info msg="task-result changed" namespace=argo nodeID=dag-test-t9l2s-4086475859 workflow=dag-test-t9l2s
time="2024-04-11T06:44:34.036Z" level=info msg="node changed" namespace=argo new.message=PodInitializing new.phase=Pending new.progress=0/1 nodeID=dag-test-t9l2s-4103253478 old.message= old.phase=Pending old.progress=0/1 workflow=dag-test-t9l2s
time="2024-04-11T06:44:34.036Z" level=info msg="node changed" namespace=argo new.message= new.phase=Running new.progress=0/1 nodeID=dag-test-t9l2s-4086475859 old.message= old.phase=Pending old.progress=0/1 workflow=dag-test-t9l2s
time="2024-04-11T06:44:34.036Z" level=warning msg="was unable to obtain the node for dag-test-t9l2s-4187141573, taskName step4"
time="2024-04-11T06:44:34.036Z" level=warning msg="was unable to obtain the node for dag-test-t9l2s-4069698240, taskName step3"
time="2024-04-11T06:44:34.036Z" level=warning msg="was unable to obtain the node for dag-test-t9l2s-4069698240, taskName step3"
time="2024-04-11T06:44:34.036Z" level=warning msg="was unable to obtain the node for dag-test-t9l2s-4187141573, taskName step4"
time="2024-04-11T06:44:34.036Z" level=warning msg="was unable to obtain the node for dag-test-t9l2s-4069698240, taskName step3"
time="2024-04-11T06:44:34.036Z" level=warning msg="was unable to obtain the node for dag-test-t9l2s-4187141573, taskName step4"
time="2024-04-11T06:44:34.036Z" level=info msg="TaskSet Reconciliation" namespace=argo workflow=dag-test-t9l2s
time="2024-04-11T06:44:34.036Z" level=info msg=reconcileAgentPod namespace=argo workflow=dag-test-t9l2s
time="2024-04-11T06:44:34.042Z" level=info msg="cleaning up pod" action=terminateContainers key=argo/dag-test-t9l2s-run-pod-2-4086475859/terminateContainers
time="2024-04-11T06:44:34.042Z" level=info msg="https://172.19.0.1:443/api/v1/namespaces/argo/pods/dag-test-t9l2s-run-pod-2-4086475859/exec?command=%2Fvar%2Frun%2Fargo%2Fargoexec&command=kill&command=15&command=1&container=wait&stderr=true&stdout=true&tty=false"
time="2024-04-11T06:44:34.046Z" level=info msg="Workflow update successful" namespace=argo phase=Running resourceVersion=15178770 workflow=dag-test-t9l2s
time="2024-04-11T06:44:34.063Z" level=info msg="signaled container" container=wait error="unable to upgrade connection: container not found (\"wait\")" namespace=argo pod=dag-test-t9l2s-run-pod-2-4086475859 stderr="<nil>" stdout="<nil>"
time="2024-04-11T06:44:44.046Z" level=info msg="Processing workflow" namespace=argo workflow=dag-test-t9l2s
time="2024-04-11T06:44:44.046Z" level=info msg="Task-result reconciliation" namespace=argo numObjs=2 workflow=dag-test-t9l2s
time="2024-04-11T06:44:44.046Z" level=info msg="task-result changed" namespace=argo nodeID=dag-test-t9l2s-4086475859 workflow=dag-test-t9l2s
time="2024-04-11T06:44:44.046Z" level=info msg="task-result changed" namespace=argo nodeID=dag-test-t9l2s-4103253478 workflow=dag-test-t9l2s
time="2024-04-11T06:44:44.047Z" level=info msg="node changed" namespace=argo new.message= new.phase=Succeeded new.progress=0/1 nodeID=dag-test-t9l2s-4103253478 old.message=PodInitializing old.phase=Pending old.progress=0/1 workflow=dag-test-t9l2s
time="2024-04-11T06:44:44.047Z" level=info msg="Pod failed: Error (exit code 1)" displayName=step2 namespace=argo pod=dag-test-t9l2s-run-pod-2-4086475859 templateName=run-pod-2 workflow=dag-test-t9l2s
time="2024-04-11T06:44:44.047Z" level=info msg="node changed" namespace=argo new.message="Error (exit code 1)" new.phase=Failed new.progress=0/1 nodeID=dag-test-t9l2s-4086475859 old.message= old.phase=Running old.progress=0/1 workflow=dag-test-t9l2s
time="2024-04-11T06:44:44.047Z" level=warning msg="was unable to obtain the node for dag-test-t9l2s-4187141573, taskName step4"
time="2024-04-11T06:44:44.047Z" level=warning msg="was unable to obtain the node for dag-test-t9l2s-4069698240, taskName step3"
time="2024-04-11T06:44:44.047Z" level=warning msg="was unable to obtain the node for dag-test-t9l2s-4069698240, taskName step3"
time="2024-04-11T06:44:44.047Z" level=warning msg="was unable to obtain the node for dag-test-t9l2s-4069698240, taskName step3"
time="2024-04-11T06:44:44.047Z" level=info msg="All of node dag-test-t9l2s.step3 dependencies [step1 step2] completed" namespace=argo workflow=dag-test-t9l2s
time="2024-04-11T06:44:44.047Z" level=warning msg="Node was nil, will be initialized as type Skipped" namespace=argo workflow=dag-test-t9l2s
time="2024-04-11T06:44:44.050Z" level=info msg="Pod node dag-test-t9l2s-4069698240 initialized Pending" namespace=argo workflow=dag-test-t9l2s
time="2024-04-11T06:44:44.072Z" level=info msg="Created pod: dag-test-t9l2s.step3 (dag-test-t9l2s-run-pod-1-4069698240)" namespace=argo workflow=dag-test-t9l2s
time="2024-04-11T06:44:44.072Z" level=warning msg="was unable to obtain the node for dag-test-t9l2s-4187141573, taskName step4"
time="2024-04-11T06:44:44.072Z" level=warning msg="was unable to obtain the node for dag-test-t9l2s-4187141573, taskName step4"
time="2024-04-11T06:44:44.072Z" level=info msg="TaskSet Reconciliation" namespace=argo workflow=dag-test-t9l2s
time="2024-04-11T06:44:44.072Z" level=info msg=reconcileAgentPod namespace=argo workflow=dag-test-t9l2s
time="2024-04-11T06:44:44.080Z" level=info msg="Workflow update successful" namespace=argo phase=Running resourceVersion=15178837 workflow=dag-test-t9l2s
time="2024-04-11T06:44:49.081Z" level=info msg="cleaning up pod" action=deletePod key=argo/dag-test-t9l2s-run-pod-1-4103253478/deletePod
time="2024-04-11T06:44:49.081Z" level=info msg="cleaning up pod" action=deletePod key=argo/dag-test-t9l2s-run-pod-2-4086475859/deletePod
time="2024-04-11T06:44:54.073Z" level=info msg="Processing workflow" namespace=argo workflow=dag-test-t9l2s
time="2024-04-11T06:44:54.074Z" level=info msg="Task-result reconciliation" namespace=argo numObjs=0 workflow=dag-test-t9l2s
time="2024-04-11T06:44:54.074Z" level=info msg="node changed" namespace=argo new.message=PodInitializing new.phase=Pending new.progress=0/1 nodeID=dag-test-t9l2s-4069698240 old.message= old.phase=Pending old.progress=0/1 workflow=dag-test-t9l2s
time="2024-04-11T06:44:54.074Z" level=warning msg="was unable to obtain the node for dag-test-t9l2s-4187141573, taskName step4"
time="2024-04-11T06:44:54.074Z" level=warning msg="was unable to obtain the node for dag-test-t9l2s-4187141573, taskName step4"
time="2024-04-11T06:44:54.074Z" level=warning msg="was unable to obtain the node for dag-test-t9l2s-4187141573, taskName step4"
time="2024-04-11T06:44:54.074Z" level=info msg="TaskSet Reconciliation" namespace=argo workflow=dag-test-t9l2s
time="2024-04-11T06:44:54.074Z" level=info msg=reconcileAgentPod namespace=argo workflow=dag-test-t9l2s
time="2024-04-11T06:44:54.083Z" level=info msg="Workflow update successful" namespace=argo phase=Running resourceVersion=15178921 workflow=dag-test-t9l2s
time="2024-04-11T06:45:04.064Z" level=info msg="cleaning up pod" action=killContainers key=argo/dag-test-t9l2s-run-pod-2-4086475859/killContainers
time="2024-04-11T06:45:11.132Z" level=info msg="Processing workflow" namespace=argo workflow=dag-test-t9l2s
time="2024-04-11T06:45:11.132Z" level=info msg="Task-result reconciliation" namespace=argo numObjs=1 workflow=dag-test-t9l2s
time="2024-04-11T06:45:11.132Z" level=info msg="task-result changed" namespace=argo nodeID=dag-test-t9l2s-4069698240 workflow=dag-test-t9l2s
time="2024-04-11T06:45:11.132Z" level=info msg="node changed" namespace=argo new.message= new.phase=Succeeded new.progress=0/1 nodeID=dag-test-t9l2s-4069698240 old.message=PodInitializing old.phase=Pending old.progress=0/1 workflow=dag-test-t9l2s
time="2024-04-11T06:45:11.132Z" level=warning msg="was unable to obtain the node for dag-test-t9l2s-4187141573, taskName step4"
time="2024-04-11T06:45:11.133Z" level=warning msg="was unable to obtain the node for dag-test-t9l2s-4187141573, taskName step4"
time="2024-04-11T06:45:11.133Z" level=warning msg="was unable to obtain the node for dag-test-t9l2s-4187141573, taskName step4"
time="2024-04-11T06:45:11.133Z" level=info msg="All of node dag-test-t9l2s.step4 dependencies [step3] completed" namespace=argo workflow=dag-test-t9l2s
time="2024-04-11T06:45:11.133Z" level=warning msg="Node was nil, will be initialized as type Skipped" namespace=argo workflow=dag-test-t9l2s
time="2024-04-11T06:45:11.136Z" level=info msg="Pod node dag-test-t9l2s-4187141573 initialized Pending" namespace=argo workflow=dag-test-t9l2s
time="2024-04-11T06:45:11.162Z" level=info msg="Created pod: dag-test-t9l2s.step4 (dag-test-t9l2s-run-pod-2-4187141573)" namespace=argo workflow=dag-test-t9l2s
time="2024-04-11T06:45:11.162Z" level=info msg="TaskSet Reconciliation" namespace=argo workflow=dag-test-t9l2s
time="2024-04-11T06:45:11.162Z" level=info msg=reconcileAgentPod namespace=argo workflow=dag-test-t9l2s
time="2024-04-11T06:45:11.170Z" level=info msg="Workflow update successful" namespace=argo phase=Running resourceVersion=15179031 workflow=dag-test-t9l2s
time="2024-04-11T06:45:16.171Z" level=info msg="cleaning up pod" action=deletePod key=argo/dag-test-t9l2s-run-pod-1-4069698240/deletePod
time="2024-04-11T06:45:21.164Z" level=info msg="Processing workflow" namespace=argo workflow=dag-test-t9l2s
time="2024-04-11T06:45:21.164Z" level=info msg="Task-result reconciliation" namespace=argo numObjs=0 workflow=dag-test-t9l2s
time="2024-04-11T06:45:21.164Z" level=info msg="node changed" namespace=argo new.message= new.phase=Running new.progress=0/1 nodeID=dag-test-t9l2s-4187141573 old.message= old.phase=Pending old.progress=0/1 workflow=dag-test-t9l2s
time="2024-04-11T06:45:21.164Z" level=info msg="TaskSet Reconciliation" namespace=argo workflow=dag-test-t9l2s
time="2024-04-11T06:45:21.164Z" level=info msg=reconcileAgentPod namespace=argo workflow=dag-test-t9l2s
time="2024-04-11T06:45:21.174Z" level=info msg="Workflow update successful" namespace=argo phase=Running resourceVersion=15179114 workflow=dag-test-t9l2s
time="2024-04-11T06:45:31.166Z" level=info msg="Processing workflow" namespace=argo workflow=dag-test-t9l2s
time="2024-04-11T06:45:31.166Z" level=info msg="Task-result reconciliation" namespace=argo numObjs=1 workflow=dag-test-t9l2s
time="2024-04-11T06:45:31.166Z" level=info msg="task-result changed" namespace=argo nodeID=dag-test-t9l2s-4187141573 workflow=dag-test-t9l2s
time="2024-04-11T06:45:31.166Z" level=info msg="Pod failed: Error (exit code 1)" displayName=step4 namespace=argo pod=dag-test-t9l2s-run-pod-2-4187141573 templateName=run-pod-2 workflow=dag-test-t9l2s
time="2024-04-11T06:45:31.166Z" level=info msg="node changed" namespace=argo new.message="Error (exit code 1)" new.phase=Failed new.progress=0/1 nodeID=dag-test-t9l2s-4187141573 old.message= old.phase=Running old.progress=0/1 workflow=dag-test-t9l2s
time="2024-04-11T06:45:31.167Z" level=info msg="Outbound nodes of dag-test-t9l2s set to [dag-test-t9l2s-4187141573]" namespace=argo workflow=dag-test-t9l2s
time="2024-04-11T06:45:31.167Z" level=info msg="node dag-test-t9l2s phase Running -> Failed" namespace=argo workflow=dag-test-t9l2s
time="2024-04-11T06:45:31.167Z" level=info msg="node dag-test-t9l2s finished: 2024-04-11 06:45:31.167053436 +0000 UTC" namespace=argo workflow=dag-test-t9l2s
time="2024-04-11T06:45:31.167Z" level=info msg="TaskSet Reconciliation" namespace=argo workflow=dag-test-t9l2s
time="2024-04-11T06:45:31.167Z" level=info msg=reconcileAgentPod namespace=argo workflow=dag-test-t9l2s
time="2024-04-11T06:45:31.167Z" level=info msg="Updated phase Running -> Failed" namespace=argo workflow=dag-test-t9l2s
time="2024-04-11T06:45:31.167Z" level=info msg="Marking workflow completed" namespace=argo workflow=dag-test-t9l2s
time="2024-04-11T06:45:31.167Z" level=info msg="Marking workflow as pending archiving" namespace=argo workflow=dag-test-t9l2s
time="2024-04-11T06:45:31.172Z" level=info msg="cleaning up pod" action=deletePod key=argo/dag-test-t9l2s-1340600742-agent/deletePod
time="2024-04-11T06:45:31.177Z" level=info msg="Workflow update successful" namespace=argo phase=Failed resourceVersion=15179175 workflow=dag-test-t9l2s
time="2024-04-11T06:45:31.184Z" level=info msg="archiving workflow" namespace=argo uid=3f3593ed-2a1f-4ec0-8445-97d8e1abe914 workflow=dag-test-t9l2s
time="2024-04-11T06:45:31.216Z" level=info msg="Queueing Failed workflow argo/dag-test-t9l2s for delete in 3s due to TTL"
time="2024-04-11T06:45:35.000Z" level=info msg="Deleting garbage collected workflow 'argo/dag-test-t9l2s'"
time="2024-04-11T06:45:35.009Z" level=info msg="Successfully request 'argo/dag-test-t9l2s' to be deleted"
time="2024-04-11T06:45:36.184Z" level=info msg="cleaning up pod" action=deletePod key=argo/dag-test-t9l2s-run-pod-2-4187141573/deletePod
time="2024-04-11T06:45:40.967Z" level=info msg="Processing workflow" namespace=argo workflow=dag-test-t9l2s
time="2024-04-11T06:45:40.968Z" level=info msg="Task-result reconciliation" namespace=argo numObjs=0 workflow=dag-test-t9l2s
time="2024-04-11T06:45:40.968Z" level=warning msg="was unable to obtain the node for dag-test-t9l2s-4187141573, taskName step4"
time="2024-04-11T06:45:40.968Z" level=warning msg="was unable to obtain the node for dag-test-t9l2s-4086475859, taskName step2"
time="2024-04-11T06:45:40.968Z" level=warning msg="was unable to obtain the node for dag-test-t9l2s-4086475859, taskName step2"
time="2024-04-11T06:45:40.968Z" level=error msg="Failed to build local scope from task" namespace=argo taskName=step3 workflow=dag-test-t9l2s
time="2024-04-11T06:45:40.968Z" level=error msg="Mark error node" error="Ancestor task node step2 not found" namespace=argo nodeName=dag-test-t9l2s.step3 workflow=dag-test-t9l2s
time="2024-04-11T06:45:40.968Z" level=error msg="node is already fulfilled" fromPhase=Succeeded namespace=argo nodeName=dag-test-t9l2s.step3 toPhase=Error workflow=dag-test-t9l2s
time="2024-04-11T06:45:40.968Z" level=info msg="node dag-test-t9l2s-4069698240 phase Succeeded -> Error" namespace=argo workflow=dag-test-t9l2s
time="2024-04-11T06:45:40.968Z" level=info msg="node dag-test-t9l2s-4069698240 message: Ancestor task node step2 not found" namespace=argo workflow=dag-test-t9l2s
time="2024-04-11T06:45:40.968Z" level=warning msg="was unable to obtain the node for dag-test-t9l2s-4187141573, taskName step4"
time="2024-04-11T06:45:40.971Z" level=info msg="Skipped node dag-test-t9l2s-4187141573 initialized Omitted (message: omitted: depends condition not met)" namespace=argo workflow=dag-test-t9l2s
time="2024-04-11T06:45:40.971Z" level=warning msg="was unable to obtain the node for dag-test-t9l2s-4086475859, taskName step2"
time="2024-04-11T06:45:40.971Z" level=warning msg="was unable to obtain the node for dag-test-t9l2s-4086475859, taskName step2"
time="2024-04-11T06:45:40.971Z" level=error msg="Mark error node" error="Ancestor task node step2 not found" namespace=argo nodeName=dag-test-t9l2s workflow=dag-test-t9l2s
time="2024-04-11T06:45:40.971Z" level=info msg="node dag-test-t9l2s phase Running -> Error" namespace=argo workflow=dag-test-t9l2s
time="2024-04-11T06:45:40.971Z" level=info msg="node dag-test-t9l2s message: Ancestor task node step2 not found" namespace=argo workflow=dag-test-t9l2s
time="2024-04-11T06:45:40.971Z" level=info msg="node dag-test-t9l2s finished: 2024-04-11 06:45:40.971781655 +0000 UTC" namespace=argo workflow=dag-test-t9l2s
time="2024-04-11T06:45:40.971Z" level=error msg="Mark error node" error="Ancestor task node step2 not found" namespace=argo nodeName=dag-test-t9l2s workflow=dag-test-t9l2s
time="2024-04-11T06:45:40.971Z" level=error msg="error in entry template execution" error="Ancestor task node step2 not found" namespace=argo workflow=dag-test-t9l2s
time="2024-04-11T06:45:40.971Z" level=info msg="Updated phase Running -> Error" namespace=argo workflow=dag-test-t9l2s
time="2024-04-11T06:45:40.971Z" level=info msg="Updated message  -> error in entry template execution: Ancestor task node step2 not found" namespace=argo workflow=dag-test-t9l2s
time="2024-04-11T06:45:40.971Z" level=info msg="Marking workflow completed" namespace=argo workflow=dag-test-t9l2s
time="2024-04-11T06:45:40.971Z" level=info msg="Marking workflow as pending archiving" namespace=argo workflow=dag-test-t9l2s
time="2024-04-11T06:45:40.977Z" level=info msg="cleaning up pod" action=deletePod key=argo/dag-test-t9l2s-1340600742-agent/deletePod
time="2024-04-11T06:45:40.979Z" level=info msg="Workflow update successful" namespace=argo phase=Error resourceVersion=15179242 workflow=dag-test-t9l2s
time="2024-04-11T06:45:40.982Z" level=info msg="archiving workflow" namespace=argo uid=16d51ef9-fd15-493b-a683-c00e1216bec8 workflow=dag-test-t9l2s
time="2024-04-11T06:45:41.014Z" level=info msg="Queueing Error workflow argo/dag-test-t9l2s for delete in 2s due to TTL"
time="2024-04-11T06:45:44.000Z" level=info msg="Deleting garbage collected workflow 'argo/dag-test-t9l2s'"
time="2024-04-11T06:45:44.010Z" level=info msg="Successfully request 'argo/dag-test-t9l2s' to be deleted"
```

### Logs from in your workflow's wait container

```text
kubectl logs -n argo -c wait -l workflows.argoproj.io/workflow=${workflow},workflow.argoproj.io/phase!=Succeeded

no
```

**Comment by shuangkun** [comment]
May look like this: https://github.com/argoproj/argo-workflows/pull/12817

**Comment by heidongxianhua** [comment]
yeah, I have cherry-pick your code, but it still got error. @shuangkun

**Comment by agilgur5** [comment]
Sounds like a duplicate of #12156?

**Comment by heidongxianhua** [comment]
add some details, when i got this error, i have searched issues and  solution in argo docs and github, including the issue https://github.com/argoproj/argo-workflows/issues/12156 , https://github.com/argoproj/argo-workflows/issues/11395 and the related prs (https://github.com/argoproj/argo-workflows/pull/12164 and https://github.com/argoproj/argo-workflows/pull/12817), i have tested localy, but this does not work. @agilgur5 @shuangkun

**Comment by shuangkun** [comment]
> yeah, I have cherry-pick your code, but it still got error. @shuangkun

could you submit a pr to fix this?

**Comment by heidongxianhua** [comment]
sorry, i have no idea to fix it now.  @shuangkun

**Comment by agilgur5** [comment]
> including the issue #12156 , #11395 and the related prs (#12164 and #12817), i have tested localy, but this does not work

For future reference, if you did that before making this issue, that would have been very helpful context to write in the issue description. Then we'd know this is not necessarily a duplicate without needing to comment 😅

**Comment by jswxstw** [comment]
> yeah, I have cherry-pick your code, but it still got error. @shuangkun

Actuall,  #12817 fixed the `argo-cli`, not the `controller`. Which way are you using to retry the failed workflow? @heidongxianhua

**Comment by agilgur5** [comment]
> Actuall, #12817 fixed the `argo-cli`, not the `controller`.

It's in the Server if I'm not mistaken -- so any method that retries it (UI, CLI, SDK, direct API) should go through the logic.
(and moving to the Controller per #12538)

**Comment by jswxstw** [comment]
> It's [in the Server](https://github.com/argoproj/argo-workflows/blob/7eaf3054bb1658e445852a7e1eecd364c50ff88a/server/workflow/workflow_server.go#L402) -- so any method that retries it (UI, CLI, SDK, direct API) should go through the logic.
> (and moving to the Controller per #12538)

Yes, I didn't express it clearly. I believe that this issue is most likely caused by using the wrong version of the CLI.

**Comment by agilgur5** [comment]
Oh did you mean _the binaries_? Since the Server is part of the CLI, i.e. `argo server`.

We refer to it as the Server. But yes, @heidongxianhua would have to cherry-pick this into their Server build and update their Server image with the build, not the Controller.

**Comment by shuangkun** [comment]
I test it in my envirment， I think it has been fixed by https://github.com/argoproj/argo-workflows/pull/12817 @heidongxianhua  can you test again use latest？

**Comment by jswxstw** [comment]
> Oh did you mean _the binaries_? Since the Server is part of the CLI, i.e. `argo server`.

😅Yes！In my opinion , CLI == _binaries_ and CLI and `argo server` should both be updated.

**Comment by agilgur5** [comment]
> 😅Yes！In my opinion , CLI == _binaries_

Generally speaking, the CLI is referred to as a client, as that is it's primary usage. A lot of people don't even know that the CLI embeds the Server (I've also seen people ask what `argo server` does as they thought it was some client command that connects to the Server or something, not that it _literally runs_ the Server).
In this case, when you referred to the CLI, that sounded like the `argo retry` command, which is not what you meant. Hence why disambiguation and specificity is important.

**Comment by heidongxianhua** [comment]
yeah, I  will have a test again now.

**Comment by heidongxianhua** [comment]
when i rebuild my argo-cli with the https://github.com/argoproj/argo-workflows/pull/12817, it works well. thank you very much. @agilgur5 @shuangkun @jswxstw

**Comment by shuangkun** [comment]
fixed by #12817
